backend/src/services/hsm/hsm-service.ts

```diff
@@ -0,0 +1,349 @@
+import grapheneLib from "graphene-pk11";
+
+import { getConfig } from "@app/lib/config/env";
+import { logger } from "@app/lib/logger";
+
+import { HsmModule, RequiredMechanisms } from "./hsm-types";
+
+type THsmServiceFactoryDep = {
+  hsmModule: HsmModule;
+};
+
+const USER_ALREADY_LOGGED_IN_ERROR = "CKR_USER_ALREADY_LOGGED_IN";
+const WRAPPED_KEY_LENGTH = 32 + 8; // AES-256 key + padding
+
+export type THsmServiceFactory = ReturnType<typeof hsmServiceFactory>;
+
+type SyncOrAsync<T> = T | Promise<T>;
+type SessionCallback<T> = (session: grapheneLib.Session) => SyncOrAsync<T>;
+
+export const withSession = async <T>(
+  { module, graphene }: HsmModule,
+  callbackWithSession: SessionCallback<T>
+): Promise<T> => {
+  const appCfg = getConfig();
+
+  let session: grapheneLib.Session | null = null;
+  try {
+    if (!module) {
+      throw new Error("PKCS#11 module is not initialized");
+    }
+
+    // Create new session
+    const slot = module.getSlots(appCfg.HSM_SLOT);
+    // eslint-disable-next-line no-bitwise
+    if (!(slot.flags & graphene.SlotFlag.TOKEN_PRESENT)) {
+      throw new Error("Slot is not initialized");
+    }
+
+    for (let i = 0; i < 10; i += 1) {
+      try {
+        // eslint-disable-next-line no-bitwise
+        session = slot.open(graphene.SessionFlag.RW_SESSION | graphene.SessionFlag.SERIAL_SESSION);
+        session.login(appCfg.HSM_PIN!);
+      } catch (error) {
+        if ((error as Error)?.message !== USER_ALREADY_LOGGED_IN_ERROR) {
+          throw error;
+        }
+        logger.warn("HSM session already logged in");
+        session = null;
+      }
+
+      if (session) {
+        break;
+      }
+
+      logger.warn("Waiting for session to be available...");
+      // eslint-disable-next-line no-await-in-loop
+      await new Promise((resolve) => {
+        let sleepAmount = 1_500 * (i + 1);
+        if (sleepAmount > 5000) sleepAmount = 5000;
+
+        setTimeout(resolve, sleepAmount);
+      });
+    }
+
+    if (!session) {
+      throw new Error("Failed to open session");
+    }
+
+    // Execute the callback and await its result (works for both sync and async)
+    const result = await callbackWithSession(session);
+    return result;
+  } finally {
+    // Clean up session if it was created
+    if (session) {
+      try {
+        session.logout();
+        session.close();
+      } catch (error) {
+        logger.error("Error cleaning up HSM session:", error);
+      }
+    }
+  }
+};
+
+// eslint-disable-next-line no-empty-pattern
+export const hsmServiceFactory = ({ hsmModule: { module, graphene } }: THsmServiceFactoryDep) => {
+  const appCfg = getConfig();
+
+  // Constants for buffer structure
+  const IV_LENGTH = 16;
```

Comment:
Can you double check if `IV_LENGTH=12` would work here? I believe NIST recommends 96-bit IV / nonces for AES-256-GCM.